src/PlusDataCollection/OpenHaptics/vtkPlusOpenHapticsDevice.cxx

```diff
@@ -0,0 +1,297 @@
+  vtkPlusDataSource* forceInput;
+  if (!inputChannels.empty() && inputChannels[0]->GetOwnerDevice()->GetToolByPortName("Force", forceInput))
+  {
+    StreamBufferItem* item = new StreamBufferItem();
+    if (forceInput->GetLatestItemUidInBuffer() > 0)
+    {
+      forceInput->GetLatestStreamBufferItem(item);
+      vtkSmartPointer<vtkMatrix4x4> forceMatrix = vtkSmartPointer<vtkMatrix4x4>::New();
+      item->GetMatrix(forceMatrix);
+      vtkMatrix4x4::Multiply4x4(rasCorrection, forceMatrix, forceMatrix);
+      force[0] = forceMatrix->GetElement(0,3);
+      force[1] = forceMatrix->GetElement(1,3);
+      force[2] = forceMatrix->GetElement(2,3);
+    }
+    delete item;
+  }
+  else
+  {
+    LOG_TRACE("No force data available");
+  }
+
+  //Current position data is pulled here
+  HDdouble trans[16];
+  HDdouble pos[3];  
+  HDdouble vel[3];
+  HDint buttonVals = 0;
+  HDboolean inkwell = 0;
+  hdBeginFrame(handle);
+  hdMakeCurrentDevice(handle);  
+  hdGetDoublev(HD_CURRENT_TRANSFORM, trans);
+  hdGetDoublev(HD_CURRENT_POSITION, pos);
+  hdGetDoublev(HD_CURRENT_VELOCITY, vel);
+  hdSetDoublev(HD_CURRENT_FORCE, force);
+  hdGetIntegerv(HD_CURRENT_BUTTONS, &buttonVals);
+  hdGetBooleanv(HD_CURRENT_INKWELL_SWITCH, &inkwell);
+  hdEndFrame(handle);
+
+  vtkPlusDataSource* stylus = NULL;
+  vtkPlusDataSource* velocity = NULL;
+  vtkPlusDataSource* buttons = NULL;
+
+  vtkSmartPointer<vtkTransform> transform = vtkSmartPointer<vtkTransform>::New();
+  vtkSmartPointer<vtkMatrix4x4> toolMatrix = vtkSmartPointer<vtkMatrix4x4>::New();
+  vtkSmartPointer<vtkTransform> rotation = vtkSmartPointer<vtkTransform>::New();
+  vtkSmartPointer<vtkTransform> velTrans = vtkSmartPointer<vtkTransform>::New();
+  vtkSmartPointer<vtkMatrix4x4> velMatrix = vtkSmartPointer<vtkMatrix4x4>::New();
+  vtkSmartPointer<vtkMatrix4x4> buttonMatrix = vtkSmartPointer<vtkMatrix4x4>::New();
+  double orient[3];
+
+  //Arrange transformations in correct order
+  transform->Identity();
+  transform->Translate(pos);
+  rotation->SetMatrix(trans);
+  rotation->GetOrientation(orient);
+  transform->RotateX(-1* orient[0] + 180);
+  transform->RotateY(orient[1]);
+  transform->RotateZ(orient[2]);  
+  transform->GetMatrix(toolMatrix);
+  velTrans->Translate(vel);
+  velTrans->GetMatrix(velMatrix);
+  vtkMatrix4x4::Multiply4x4(rasCorrection, velMatrix, velMatrix);
+  vtkMatrix4x4::Multiply4x4(rasCorrection,toolMatrix,toolMatrix);
+
+
+  buttonMatrix->SetElement(0,0,0);
+  buttonMatrix->SetElement(1, 1, 0);
+  buttonMatrix->SetElement(2, 2, 0);
+  buttonMatrix->SetElement(0, 0, (bool)(buttonVals & HD_DEVICE_BUTTON_1));
+  buttonMatrix->SetElement(1, 0, (bool)(buttonVals & HD_DEVICE_BUTTON_2));
+  buttonMatrix->SetElement(2, 0, (bool)(buttonVals & HD_DEVICE_BUTTON_3));
+  buttonMatrix->SetElement(3, 0, (bool)(buttonVals & HD_DEVICE_BUTTON_4));
+  buttonMatrix->SetElement(0, 1, (int)inkwell);
+
+
+  if (client->GetToolByPortName("Stylus", stylus) == PLUS_SUCCESS)

```

Comment:
You can simplify this a bit by just calling 
```
if(ToolTimeStampedUpdate("Stylus", ...) != PLUS_SUCCESS)
{
  LOG_ERROR(...)
}
```

Reply:
So, this accesses the tools by ID instead of portname, which does not seem to be working for me (can't find any of the tools based on their names in the config).  Same with client->InputChannels[0].GetTool(forceInput, "Force") you suggested above.  I have the IDs set in the configuration the same as the port name, but that internal ID seems to be something different?

Reply:
The ID that ToolTimeStampedUpdate is expecting is the transform name id (ex. StylusToTracker, etc).

Reply:
Right, yes. Nevermind.